text/0000-standard-lazy-types.md

```diff
@@ -0,0 +1,416 @@
+
+Similarly, because `OnceCell` provides strong exception safety guarantee, it implements `UnwindSafe`:
+
+```rust
+impl<T: UnwindSafe>                    UnwindSafe for OnceCell<T> {}
+impl<T: UnwindSafe + RefUnwindSafe> RefUnwindSafe for OnceCell<T> {}
+```
+
+## Default type parameter on `Lazy`
+
+`Lazy` is defined with default type parameter.
+
+```rust
+pub struct Lazy<T, F = fn() -> T> { ... }
+```
+
+This is important to make using `Lazy` in static contexts convenient.
+Without this default, the user would have to type `T` type twice:
+
+```rust
+static GLOBAL_DATA: Lazy<Mutex<HashMap<i32, String>>, fn() -> Mutex<HashMap<i32, String>>
+    = Lazy::new(|| ... );
+```
+
+If we allow type inference in statics, this could be shortened to
+
+```rust
+static GLOBAL_DATA: Lazy<Mutex<HashMap<i32, String>>, _>
+    = Lazy::new(|| ... );
+```
+
+There are two drawbacks of using fn pointer type:
+
+* fn pointers are not ZSTs, so we waste one pointer per static lazy value.
+  Lazy locals will generally rely on type-inference and will use more specific closure type.
+* Specifying type for local lazy value might be tricky: `let x: Lazy<i32> = Lazy::new(|| closed_over_var)` fails with type error, the correct syntax is `let x: Lazy<i32, _> = Lazy::new(|| closed_over_var)`.
+
+## Only thread-safe flavor
+
+It is possible to add only `sync` version of the types, as they are the most useful.
+However, this will be against zero cost abstractions spirit.
+Additionally, non thread-safe version is required to replace `thread_local!` macro without imposing synchronization.
+
+# Prior art
+[prior-art]: #prior-art
+
+The primary bit of prior art here is the [`once_cell`] library, which itself draws on multiple sources:
+
+* [double-checked-cell](https://crates.io/crates/double-checked-cell)
+* [lazy-init](https://crates.io/crates/lazy-init)
+* [lazycell](https://crates.io/crates/lazycell)
+* [mitochondria](https://crates.io/crates/mitochondria)
+* [lazy_static](https://crates.io/crates/lazy_static)
+
+Many languages provide library-defined lazy values, for example [Kotlin](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin/lazy.html#kotlin$lazy(kotlin.Function0((kotlin.lazy.T)))).
+Typically, a lazy value is just a wrapper around closure.
+This design doesn't always work in Rust, as closing over `self` runs afoul of the borrow checker, we need a more primitive `OnceCell` type.
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+- What is the best naming/place for these types?
+- What is the best naming scheme for methods? Is it `get_or_try_init` or `try_inert_with`?
+- Is the `F = fn() -> T` hack worth it?
+
+# Future possibilities
+[future-possibilities]: #future-possibilities
+
+* Once `#[thread_local]` attribute is stable, `cell::Lazy` can serve as a replacement for `std::thread_local!` macro.
+* Supporting type inference in constants might allow us to drop the default type parameter on `Lazy`.
```

Comment:
`impl Trait` would be enough?

Reply:
Nope, spelling `impl Fn() -> T` is not better than spelling `fn() -> T`. The problem is not that fn-pointer is inneficient, the problem is that the syntax to write it is verbose.